File: examples/c-sharp-examples/FreeFormDeformation.dox.cs
public class FreeFormDeformationExample
{
    public static void Main(string[] args)
    {
        try
        {
            // Make a mesh to deform
            // (to deform your own mesh, load it with MR.MeshLoad.fromAnySupportedFormat("mesh.stl") instead)
            var mesh = MR.makeUVSphere(1.0f, 32, 32);

            // Compute mesh bounding box
            MR.Box3f box = mesh.getBoundingBox();

            // Construct deformer on mesh vertices
            MR.FreeFormDeformer ffDeformer = new(mesh);

            // Init deformer with 3x3 grid on mesh box
            ffDeformer.init(MR.Vector3i.diagonal(3), box);

            // Move some control points of the grid to the center
            ffDeformer.setRefGridPointPosition(new MR.Vector3i(1, 1, 0), box.center());
            ffDeformer.setRefGridPointPosition(new MR.Vector3i(1, 1, 2), box.center());
            ffDeformer.setRefGridPointPosition(new MR.Vector3i(0, 1, 1), box.center());
            ffDeformer.setRefGridPointPosition(new MR.Vector3i(2, 1, 1), box.center());
            ffDeformer.setRefGridPointPosition(new MR.Vector3i(1, 0, 1), box.center());
            ffDeformer.setRefGridPointPosition(new MR.Vector3i(1, 2, 1), box.center());

            // Apply the deformation to the mesh vertices
            ffDeformer.apply();

            // Invalidate the mesh because of external vertex changes
            mesh.invalidateCaches();

            // Save deformed mesh
            MR.MeshSave.toAnySupportedFormat(mesh, "deformed_mesh.stl");
        }
        catch (Exception e)
        {
            Console.WriteLine("Error: {0}", e.Message);
        }
    }
}

File: examples/js-examples/FreeFormDeformation.dox.js
import createMeshLib from '@meshinspector/meshlib';

const ml = await createMeshLib();

// Load mesh
using mesh = ml.MeshLoad.fromAnySupportedFormat('mesh.stl');

// Construct the deformer on the mesh (the JS binding takes the mesh directly)
using ffDeformer = new ml.FreeFormDeformer(mesh);

// Compute mesh bounding box
using box = mesh.computeBoundingBox();

// Init deformer with a 3x3x3 grid on the mesh box
ffDeformer.init({ x: 3, y: 3, z: 3 }, box);

// Move some control points of the grid to the center
const center = box.center();
ffDeformer.setRefGridPointPosition({ x: 1, y: 1, z: 0 }, center);
ffDeformer.setRefGridPointPosition({ x: 1, y: 1, z: 2 }, center);
ffDeformer.setRefGridPointPosition({ x: 0, y: 1, z: 1 }, center);
ffDeformer.setRefGridPointPosition({ x: 2, y: 1, z: 1 }, center);
ffDeformer.setRefGridPointPosition({ x: 1, y: 0, z: 1 }, center);
ffDeformer.setRefGridPointPosition({ x: 1, y: 2, z: 1 }, center);

// Apply the deformation to the mesh vertices
ffDeformer.apply();

// Invalidate the mesh because of the external vertex changes
mesh.invalidateCaches();

// Save deformed mesh
ml.MeshSave.toAnySupportedFormat(mesh, 'deformed_mesh.stl');
